### PR TITLE
Avoid blocking ThreadPool threads in cancelable NamedPipe*Stream.ReadAsync on Unix

### PR DESCRIPTION
**Pull request by stephentoub**
Named pipes on Unix are implemented via sockets.  When a ReadAsync is issued, that it turns results in a ReceiveAsync on the socket.  But today we lack a ReceiveAsync overload that takes a CancellationToken, and until such an overload exists, we need a workaround.

Today's workaround involves polling and blocking a thread pool thread doing a synchronous read; that's very expensive, especially when many such calls are made across one or more pipes and we end up blocking many threads.

This PR provides an alternative workaround.  Instead of queueing a blocking poll to the thread pool, we issue a zero-length async receive so that we're notified when data is available, and we concurrently monitor the cancellation token.  This is more efficient when cancellation doesn't occur.  If cancellation does occur, we potentially "leak" the read request until data is available (or the socket closes), but cancellation is relatively rare, and even more rare is reusing the stream after cancellation has occurred, so this shouldn't be much more than a small fixed overhead.

cc: @geoffkizer, @JeremyKuhne, @pjanotti, @wfurt 
Fixes https://github.com/dotnet/corefx/issues/24095

### CODE REVIEW

File: src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs [hunk 1]
```diff
@@ -162,38 +162,48 @@ private unsafe void WriteCore(ReadOnlySpan<byte> buffer)
...
-            try
-            {
-                if (destination.TryGetArray(out ArraySegment<byte> buffer))
-                {
-                    return await socket.ReceiveAsync(buffer, SocketFlags.None).ConfigureAwait(false);
-                }
-                else
-                {
-                    // TODO #22608: Remove this terribly inefficient special-case once Socket.ReceiveAsync
-                    // accepts a Memory<T> in the near future.
```
Comment (pjanotti) [inline src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs]: Q.: since #22608 is still open shouldn't this comment still be kept as a comment?

Comment (stephentoub) [reply src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs]: I effectively just moved it up to here:
https://github.com/dotnet/corefx/pull/24302/files#diff-3aff29b58d0a45aaaba7664306a8a145R167

Comment (pjanotti) [reply src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs]: Somehow I didn't see it. LGTM.

---
In reply to: [141642990](https://github.com/dotnet/corefx/pull/24302#discussion_r141642990) [](ancestors = 141642990)

**Comment by stephentoub** [comment]
@ericstj, I've been seeing this error a bunch in CI over the last few days:
https://ci3.dot.net/job/dotnet_corefx/job/master/job/windows-TGroup_all+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/3676/
```
D:\j\workspace\windows-TGrou---0d2c9ac4\Tools\Packaging.targets(1068,5): error MSB4018: The "GeneratePackageReport" task failed unexpectedly. [D:\j\workspace\windows-TGrou---0d2c9ac4\pkg\Microsoft.NETCore.Targets\Microsoft.NETCore.Targets.pkgproj]
D:\j\workspace\windows-TGrou---0d2c9ac4\Tools\Packaging.targets(1068,5): error MSB4018: System.IO.IOException: The process cannot access the file 'D:\j\workspace\windows-TGrou---0d2c9ac4\pkg\Microsoft.NETCore.Platforms\runtime.json' because it is being used by another process. [D:\j\workspace\windows-TGrou---0d2c9ac4\pkg\Microsoft.NETCore.Targets\Microsoft.NETCore.Targets.pkgproj]
D:\j\workspace\windows-TGrou---0d2c9ac4\Tools\Packaging.targets(1068,5): error MSB4018:    at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath) [D:\j\workspace\windows-TGrou---0d2c9ac4\pkg\Microsoft.NETCore.Targets\Microsoft.NETCore.Targets.pkgproj]
D:\j\workspace\windows-TGrou---0d2c9ac4\Tools\Packaging.targets(1068,5): error MSB4018:    at System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost) [D:\j\workspace\windows-TGrou---0d2c9ac4\pkg\Microsoft.NETCore.Targets\Microsoft.NETCore.Targets.pkgproj]
D:\j\workspace\windows-TGrou---0d2c9ac4\Tools\Packaging.targets(1068,5): error MSB4018:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share) [D:\j\workspace\windows-TGrou---0d2c9ac4\pkg\Microsoft.NETCore.Targets\Microsoft.NETCore.Targets.pkgproj]
D:\j\workspace\windows-TGrou---0d2c9ac4\Tools\Packaging.targets(1068,5): error MSB4018:    at NuGet.RuntimeModel.JsonRuntimeFormat.ReadRuntimeGraph(String filePath) [D:\j\workspace\windows-TGrou---0d2c9ac4\pkg\Microsoft.NETCore.Targets\Microsoft.NETCore.Targets.pkgproj]
D:\j\workspace\windows-TGrou---0d2c9ac4\Tools\Packaging.targets(1068,5): error MSB4018:    at Microsoft.DotNet.Build.Tasks.Packaging.AggregateNuGetAssetResolver..ctor(String runtimeFile) [D:\j\workspace\windows-TGrou---0d2c9ac4\pkg\Microsoft.NETCore.Targets\Microsoft.NETCore.Targets.pkgproj]
D:\j\workspace\windows-TGrou---0d2c9ac4\Tools\Packaging.targets(1068,5): error MSB4018:    at Microsoft.DotNet.Build.Tasks.Packaging.GeneratePackageReport.LoadFiles() [D:\j\workspace\windows-TGrou---0d2c9ac4\pkg\Microsoft.NETCore.Targets\Microsoft.NETCore.Targets.pkgproj]
D:\j\workspace\windows-TGrou---0d2c9ac4\Tools\Packaging.targets(1068,5): error MSB4018:    at Microsoft.DotNet.Build.Tasks.Packaging.GeneratePackageReport.Execute() [D:\j\workspace\windows-TGrou---0d2c9ac4\pkg\Microsoft.NETCore.Targets\Microsoft.NETCore.Targets.pkgproj]
D:\j\workspace\windows-TGrou---0d2c9ac4\Tools\Packaging.targets(1068,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [D:\j\workspace\windows-TGrou---0d2c9ac4\pkg\Microsoft.NETCore.Targets\Microsoft.NETCore.Targets.pkgproj]
D:\j\workspace\windows-TGrou---0d2c9ac4\Tools\Packaging.targets(1068,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext() [D:\j\workspace\windows-TGrou---0d2c9ac4\pkg\Microsoft.NETCore.Targets\Microsoft.NETCore.Targets.pkgproj]
```

**Comment by ericstj** [comment]
Thanks @stephentoub I'll look into it

**Comment by stephentoub** [comment]
Thanks, @ericstj.

**Comment by stephentoub** [comment]
@dotnet-bot test Packaging All Configurations x64 Debug Build please

**Comment by stephentoub** [comment]
@dotnet-bot test Packaging All Configurations x64 Debug Build please (same GeneratePackageReport error)

**Comment by ericstj** [comment]
That error should be fixed now, I disabled the target that writes to the file during build.

**Comment by ericstj** [comment]
CI has actually passed here, its just that GH isn't reflecting the result.  
image
@stephentoub I think you can merge this.

**Comment by geoffkizer** [comment]
LGTM

**Comment by stephentoub** [comment]
Thanks, @ericstj, and for the reviews @pjanotti and @geoffkizer.

**Pull request merged**